### PR TITLE
Export correct yaml file

### PR DESCRIPTION
**Pull request by StrongMonkey**
Add export-yml api support for ip, ip6, netalias and isolation. These options won't be added to agent until they are verified work.

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/6349

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/6281

**Comment by StrongMonkey** [comment]
`cpuSetMems`, `kernelMemory `, `sysctls ` are removed. `device_*` and `ulimits` are export as map. Tmpfs is exported as array. @joshwget @deniseschannon

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/service/impl/ServiceDiscoveryApiServiceImpl.java [hunk 1]
```diff
@@ -271,6 +275,95 @@ private void populateLogConfig(Map<String, Object> cattleServiceData, Map<String
...
+    private void populateTmpfs(Map<String, Object> cattleServiceData, Map<String, Object> composeServiceData) {
+        Object value = cattleServiceData.get(ServiceConstants.FIELD_TMPFS);
+        if (value instanceof Map) {
+            if (!((Map<?, ?>) value).isEmpty()) {
+                Map<String, Object> map = (Map<String, Object>)value;
+                Iterator<String> it = map.keySet().iterator();
+                ArrayList<String> list = new ArrayList<>();
+                while (it.hasNext()) {
+                    String key = it.next();
+                    list.add(key);
```
Comment (joshwget) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/service/impl/ServiceDiscoveryApiServiceImpl.java]: Shouldn't it be `key:value` for cases where `value != ""`?

**Comment by StrongMonkey** [comment]
`ip`, `ip6`, and `netalias` are also removed

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java [hunk 2]
```diff
@@ -13,70 +13,92 @@
...
-    public static final ServiceDiscoveryConfigItem RESTART = new ServiceDiscoveryConfigItem("restartPolicy", "restart", false);
-    public static final ServiceDiscoveryConfigItem STDINOPEN = new ServiceDiscoveryConfigItem("stdinOpen", "stdin_open", false);
+    public static final ServiceDiscoveryConfigItem HOSTNAME = new ServiceDiscoveryConfigItem("hostname", "hostname",
+            false);
+    public static final ServiceDiscoveryConfigItem DOMAINNAME = new ServiceDiscoveryConfigItem("domainName",
+            "domainname", false);
+    public static final ServiceDiscoveryConfigItem MEMLIMIT = new ServiceDiscoveryConfigItem("memory", "mem_limit",
+            false);
+    public static final ServiceDiscoveryConfigItem MEMRESERVATION = new ServiceDiscoveryConfigItem("memoryReservation",
+            "mem_reservation", false);
```
Comment (alena1108) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: is it supported by docker-compose already? I see 2 conflicting PRs for this parameter: one named "mem_reservation" and another is "memory_reservation", could you double check and provide a link to a finite parameter?

https://github.com/docker/compose/pull/3714/files
https://github.com/docker/compose/pull/3723/files

File: tests/integration-v1/cattletest/core/test_svc_discovery.py [hunk 3]
```diff
@@ -1835,12 +1835,37 @@ def test_export_config(client, context):
...
+                         }
+                     },
+                     "tmpfs": {"/run": "rw"},
+                     "ulimits": [{"name": "cpu", "soft": 1234, "hard": 1234},
+                                 {"name": "nporc", "soft": 1234}]
+                     }
     service = client. \
         create_service(name="web",
-                       environmentId=env.id,
+                       stackId=env.id,
```
Comment (alena1108) [inline tests/integration-v1/cattletest/core/test_svc_discovery.py]: it is an environmentId in v1 and stackId in v2, no need to change

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java [hunk 4]
```diff
@@ -13,70 +13,92 @@
...
     public static final ServiceDiscoveryConfigItem DNSOPT = new ServiceDiscoveryConfigItem("dnsOpt", "dns_opt", false);
-    public static final ServiceDiscoveryConfigItem GROUPADD = new ServiceDiscoveryConfigItem("groupAdd", "groud_add", false);
-    public static final ServiceDiscoveryConfigItem KERNELMEMORY = new ServiceDiscoveryConfigItem("kernelMemory", "kernel_memory", false);
-    public static final ServiceDiscoveryConfigItem MEMORYSWAPPINESS = new ServiceDiscoveryConfigItem("memorySwappiness", "memory_swappiness", false);
-    public static final ServiceDiscoveryConfigItem OOMKILLDISABLE = new ServiceDiscoveryConfigItem("oomKillDisable", "oom_kill_disable", false);
-    public static final ServiceDiscoveryConfigItem SHMSIZE = new ServiceDiscoveryConfigItem("shmSize", "shm_size", false);
-    public static final ServiceDiscoveryConfigItem TMPFS = new ServiceDiscoveryConfigItem("tmpfs", "tmpfs", false);
+    public static final ServiceDiscoveryConfigItem GROUPADD = new ServiceDiscoveryConfigItem("groupAdd", "group_add",
+            false);
+    public static final ServiceDiscoveryConfigItem EXTRAHOSTS = new ServiceDiscoveryConfigItem("extraHosts", "extra_hosts", false);
```
Comment (alena1108) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: can we add a test for extra_hosts?

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java [hunk 5]
```diff
@@ -13,70 +13,92 @@
...
+    public static final ServiceDiscoveryConfigItem SECURITYOPT = new ServiceDiscoveryConfigItem("securityOpt", "security_opt", false);
+    public static final ServiceDiscoveryConfigItem READONLY = new ServiceDiscoveryConfigItem("readOnly", "readOnly", false);
+    public static final ServiceDiscoveryConfigItem MEMORYSWAPPINESS = new ServiceDiscoveryConfigItem("memorySwappiness",
+            "mem_swappiness", false);
+    public static final ServiceDiscoveryConfigItem OOMKILLDISABLE = new ServiceDiscoveryConfigItem("oomKillDisable",
+            "oom_kill_disable", false);
```
Comment (alena1108) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: test for oom_kill_disable :)

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java [hunk 6]
```diff
@@ -13,70 +13,92 @@
...
-    public static final ServiceDiscoveryConfigItem MILLICPURESERVATION = new ServiceDiscoveryConfigItem("milliCpuReservation", "milli_cpu_reservation", true,
-            false, false);
-    public static final ServiceDiscoveryConfigItem VOLUME_DRIVER = new ServiceDiscoveryConfigItem("volumeDriver", "volume_driver", false);
+    public static final ServiceDiscoveryConfigItem STOPSIGNAL = new ServiceDiscoveryConfigItem("stopSignal",
+            "stop_signal", false);
+    public static final ServiceDiscoveryConfigItem OOMSCOREADJ = new ServiceDiscoveryConfigItem("oomScoreAdj",
+            "oom_score_adj", false);
+    public static final ServiceDiscoveryConfigItem IPCMODE = new ServiceDiscoveryConfigItem("ipcMode", "ipc_mode",
+            false);
+    public static final ServiceDiscoveryConfigItem ISOLATION = new ServiceDiscoveryConfigItem("isolation", "isolation",
```
Comment (alena1108) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: test for isolation is missing

File: code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java (hunk 2 above)
Comment (StrongMonkey) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: @joshwget verified that it is `mem_reservation`

Comment (joshwget) [inline code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/resource/ServiceDiscoveryConfigItem.java]: It's not supported yet, but the name `mem_reservation` seems decided on. I think the first instance of `memory_reservation` might actually be a mistake in the PR.

**Comment by StrongMonkey** [comment]
@alena1108 missing test cases added.

File: tests/integration-v1/cattletest/core/test_svc_discovery.py (hunk 3 above)
Comment (alena1108) [inline tests/integration-v1/cattletest/core/test_svc_discovery.py]: @StrongMonkey could you change it back to environmentId?

**Comment by alena1108** [comment]
@StrongMonkey could you fix v1 test by adding environmentId back? stack used to be called an environment in v1

File: tests/integration-v1/cattletest/core/test_svc_discovery.py [hunk 7]
```diff
@@ -1922,7 +1945,7 @@ def test_export_config(client, context):
                                  "restartPolicy": restart_policy}
     service_nolog = client. \
         create_service(name="web-nolog",
-                       stackId=env.id,
+                       environmentId=env.id,
```
Comment (StrongMonkey) [inline tests/integration-v1/cattletest/core/test_svc_discovery.py]: @alena1108 i fixed it in here

**Comment by alena1108** [comment]
LGTM

**Pull request merged**